### DISCUSSION TITLE
Broadcasting

### DISCUSSION THREAD

**Discussion by yovanoc** [uNetworking/uWebSockets.js#140] [Q&A] [answered]
Hello, I'm using uWebSockets.js along with a Protobuf protocol for a production project. It's really amazing! But is there a way to broadcast data to all currently connected clients out of the box? or the best way to do it, thanks

**Comment by ghost** [comment]
If uWebsocket.js doesn't collect all connections in array or set, you should do it yourself and then, by iterating through this array send message to each of stored connection. It's how broadcasting works.

**Comment by yovanoc** [comment]
Yes I tried that but the « ws » in the close event is just an empty object so I don’t know how to keep track and correctly remove clients when they disconnects

**Comment by dbl0null** [comment]
might be related: https://github.com/uNetworking/uWebSockets/issues/875
https://github.com/uNetworking/uWebSockets/issues/790
https://github.com/uNetworking/uWebSockets.js/issues/57

**Comment by yovanoc** [comment]
Yes with pubsub we can do something like broadcasting but what if we want to keep track of connected users and when they disconnect to keep these events?

**Comment by ghost** [comment]
It's kind of weird that we can only publish by ws object, when it shouldn't be related. Publishing should be a thing of app or sth like wsManager inside app.

https://github.com/uNetworking/uWebSockets.js/blob/master/examples/PubSub.js

backing to question: ws should not be undefined when close event occurs, it has to be bug I guess

**Comment by dbl0null** [comment]
also there's currently no way to do "frame-once/send-many" (discussed somewhere in https://github.com/uNetworking/uWebSockets/issues)

so the only thing we can do for now is to track active connections manually
personally i'm using two Sets: one for "active" connections, and one for "idle" connections
so emulating broadcast (pub/sub with url client connected to as **group**) would be like:

```javascript
const bufferToString = buf => new Uint8Array(buf).join('.')
const getIpAddress = (res, req) => (req && req.getHeader('x-forwarded-for')) || bufferToString(res.getRemoteAddress())

const main = uws.App({})

const clients = new Set()
const idle = new Set()

main.ws('*', {
  compression: 0,
  maxPayloadLength: 1e6,
  idleTimeout: 0,
  open: (client, req) => {
    client.endpoint = req.getUrl()
    client.ip = getIpAddress(client, req)
    clients.add(client)
  },
  message: (source, message) => {
    clients.forEach(client => {
      if (source.endpoint === client.endpoint) {
        if (!client.send(message, true, false)) {
          clients.delete(client)
          idle.add(client)
        }
      }
    })
  },
  drain: client => {
    // actually it's better to check bufferedAmount here
    clients.add(client)
    idle.delete(client)
  },
  close: client => {
    clients.delete(client)
    idle.delete(client)
  }
})
```

it works 
you know "waaay better than anything available out there"

**Comment by yovanoc** [comment]
@Kamil93 It's not undefined but look

<img width="640" alt="Capture d’écran 2019-05-10 à 11 35 17" src="https://user-images.githubusercontent.com/7302317/57517799-ea11fc00-7317-11e9-83bb-7f7190ad255a.png">
<img width="640" alt="Capture d’écran 2019-05-10 à 11 35 17" src="https://user-images.githubusercontent.com/7302317/57517800-eaaa9280-7317-11e9-8688-3541e49ef2e5.png">


I can't call any method on this "ws"

**Comment by ghost** [comment]
@yovanoc show sth more, bit of code that displays it?

**Comment by yovanoc** [comment]
I've put the console.log in the second screen? what do you mean by "more code" ?

**Comment by dbl0null** [comment]
@yovanoc yes, you can not call any **ws** method, but it is same **ws** object you've received in **.open** handler. so you can track it

**Comment by ghost** [comment]
@yovanoc nvm, I saw there 1 screen, now are 2

as @dbl0null said, use for example Set object to track ws references, then you can use set.size to count them for example

**Comment by yovanoc** [comment]
<img width="621" alt="Capture d’écran 2019-05-10 à 11 40 56" src="https://user-images.githubusercontent.com/7302317/57518097-96ec7900-7318-11e9-874e-41ae27ee3760.png">
<img width="621" alt="Capture d’écran 2019-05-10 à 11 41 01" src="https://user-images.githubusercontent.com/7302317/57518098-96ec7900-7318-11e9-9825-4091c1647280.png">

@dbl0null Oh yes I know what you mean now, I will try that

**Comment by yovanoc** [comment]
@dbl0null Your custom forEach is to broadcast to all users connected to the same endpoint?

**Comment by ghost** [comment]
@yovanoc 

Just use:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach

**Comment by dbl0null** [comment]
@yovanoc
> forEach

drop-in replacement would be:

```javascript
    clients.forEach(client => {
      if (source.endpoint === client.endpoint) {
        if (!client.send(message, true, false)) {
          clients.delete(client)
          idle.add(client)
        }
      }
    })
```

*will update previous message

**Comment by ghost** [comment, marked as answer]
Broadcasting with a loop is the "bubble sort" solution, it works but is the least efficient way. The built in pub sub support, still unfinished, will be the way to go.

**Comment by coder456** [comment]
How much of a performance increase will there be? currently I do
```
open:(ws)=> {ws.id = ++socketCount; allSockets[ws.id] = ws},
close:(ws)=> {delete allSockets[ws.id]}
// broadcast to all
for(let s in allSockets) allSockets[s].send(message)
```
